File: disk_monitor.php
<?php
// disk_monitor.php — PHP версия

class DiskMonitor {
    private $interval;
    private $duration;
    private $historyFile;
    private $history = [];

    public function __construct($interval = 2, $duration = 60, $historyFile = 'disk_history.json') {
        $this->interval = $interval;
        $this->duration = $duration;
        $this->historyFile = $historyFile;
    }

    private function getDiskIO() {
        $result = [];
        if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN') {
            // Windows
            $result['C:'] = ['read_bytes' => 0, 'write_bytes' => 0];
        } else {
            // Linux/macOS: используем iostat
            $output = shell_exec('iostat -d 1 2 2>/dev/null');
            $lines = explode("\n", $output);
            foreach ($lines as $line) {
                $parts = preg_split('/\s+/', trim($line));
                if (count($parts) >= 7 && is_numeric($parts[5])) {
                    $disk = $parts[0];
                    $result[$disk] = [
                        'read_bytes' => (int)($parts[5] * 1024),
                        'write_bytes' => (int)($parts[6] * 1024)
                    ];
                }
            }
        }
        return $result;
    }

    private function calculateSpeeds($prev, $curr) {
        $speeds = [];
        foreach ($curr as $disk => $data) {
            if (isset($prev[$disk])) {
                $readMB = ($data['read_bytes'] - $prev[$disk]['read_bytes']) / (1024 * 1024);
                $speeds[$disk] = $readMB / $this->interval;
            }
        }
        return $speeds;
    }

    public function collect() {
        echo "\033[36m💾 Disk Monitor (History) (PHP)\033[0m\n";
        echo "Интервал: {$this->interval} сек, Длительность: {$this->duration} сек\n\n";

        $totalSamples = (int)($this->duration / $this->interval);
        echo "📊 Мониторинг дисков...\n";

        $prev = $this->getDiskIO();
        sleep($this->interval);

        for ($i = 0; $i < $totalSamples; $i++) {
            $curr = $this->getDiskIO();
            $speeds = $this->calculateSpeeds($prev, $curr);

            $entry = [
                'timestamp' => date('c'),
                'speeds' => $speeds
            ];
            $this->history[] = $entry;

            // Прогресс
            $percent = ($i + 1) / $totalSamples * 100;
            $barLen = 30;
            $filled = (int)($barLen * ($i + 1) / $totalSamples);
            $bar = str_repeat('█', $filled) . str_repeat('░', $barLen - $filled);
            fprintf(STDERR, "\r[%s] %.1f%% (%d/%d)", $bar, $percent, $i + 1, $totalSamples);

            $prev = $curr;
            sleep($this->interval);
        }

        fprintf(STDERR, "\n");
        $this->saveHistory();
        $this->printStats();
    }

    private function saveHistory() {
        $data = [
            'interval' => $this->interval,
            'duration' => $this->duration,
            'start_time' => $this->history[0]['timestamp'] ?? null,
            'end_time' => $this->history[count($this->history)-1]['timestamp'] ?? null,
            'samples' => $this->history
        ];
        file_put_contents($this->historyFile, json_encode($data, JSON_PRETTY_PRINT));
        echo "\033[32m💾 История сохранена: {$this->historyFile}\033[0m\n";
    }

    private function printStats() {
        if (empty($this->history)) return;

        $stats = [];
        foreach ($this->history as $entry) {
            foreach ($entry['speeds'] as $disk => $speed) {
                if (!isset($stats[$disk])) $stats[$disk] = [];
                $stats[$disk][] = $speed;
            }
        }

        echo "\n\033[36m📈 Статистика за период:\033[0m\n";
        echo "┌" . str_repeat("─", 80) . "┐\n";
        echo "│ " . str_pad("Диск", 10) . " " . str_pad("Среднее (МБ/с)", 15) . " " . str_pad("Макс (МБ/с)", 12) . " " . str_pad("Мин (МБ/с)", 12) . " " . str_pad("IOPS", 15) . " │\n";
        echo "├" . str_repeat("─", 80) . "┤\n";

        foreach ($stats as $disk => $speeds) {
            if (!empty($speeds)) {
                $avg = array_sum($speeds) / count($speeds);
                $max = max($speeds);
                $min = min($speeds);
                $avgIOPS = $avg;
                echo "│ " . str_pad($disk, 10) . " " . str_pad(number_format($avg, 1), 15, " ", STR_PAD_LEFT) . " " . str_pad(number_format($max, 1), 12, " ", STR_PAD_LEFT) . " " . str_pad(number_format($min, 1), 12, " ", STR_PAD_LEFT) . " " . str_pad(number_format($avgIOPS, 0), 15, " ", STR_PAD_LEFT) . " │\n";
            }
        }

        echo "└" . str_repeat("─", 80) . "┘\n";
    }

    private function loadHistory() {
        if (!file_exists($this->historyFile)) {
            echo "\033[31m❌ Файл {$this->historyFile} не найден.\033[0m\n";
            return;
        }

        $data = json_decode(file_get_contents($this->historyFile), true);
        echo "\033[36m📂 Загружена история из {$this->historyFile}\033[0m\n";
        echo "  Интервал: {$data['interval']} сек\n";
        echo "  Начало: {$data['start_time']}\n";
        echo "  Конец: {$data['end_time']}\n";
        echo "  Сэмплов: " . count($data['samples']) . "\n";
    }
}

function main($argv) {
    $interval = 2;
    $duration = 60;
    $historyFile = 'disk_history.json';
    $view = false;

    for ($i = 1; $i < count($argv); $i++) {
        if ($argv[$i] == '--interval') $interval = (int)$argv[++$i];
        else if ($argv[$i] == '--duration') $duration = (int)$argv[++$i];
        else if ($argv[$i] == '--file') $historyFile = $argv[++$i];
        else if ($argv[$i] == '--view') $view = true;
    }

    $monitor = new DiskMonitor($interval, $duration, $historyFile);

    if ($view) {
        $monitor->loadHistory();
    } else {
        $monitor->collect();
    }
}

$argc = $_SERVER['argc'] ?? 0;
$argv = $_SERVER['argv'] ?? [];
main($argv);
?>

File: disk_monitor.rb
# disk_monitor.rb — Ruby версия

require 'json'
require 'time'
require 'open3'
require 'optparse'

class DiskMonitor
  def initialize(interval: 2, duration: 60, history_file: 'disk_history.json')
    @interval = interval
    @duration = duration
    @history_file = history_file
    @history = []
  end

  def get_disk_io
    result = {}
    if RUBY_PLATFORM =~ /linux/
      # Linux: читаем /proc/diskstats
      File.readlines('/proc/diskstats').each do |line|
        parts = line.strip.split
        next if parts.length < 14
        disk = parts[2]
        read_sectors = parts[5].to_i * 512
        write_sectors = parts[9].to_i * 512
        result[disk] = { read_bytes: read_sectors, write_bytes: write_sectors }
      end
    elsif RUBY_PLATFORM =~ /darwin/
      # macOS: используем iostat
      output, _ = Open3.capture2('iostat -d 1 2 2>/dev/null')
      output.each_line do |line|
        parts = line.strip.split
        next if parts.length < 4 || parts[0] == 'disk'
        disk = parts[0]
        result[disk] = { read_bytes: (parts[3].to_f * 1024).to_i, write_bytes: (parts[4].to_f * 1024).to_i }
      end
    else
      # Windows: заглушка
      ['C:', 'D:'].each do |disk|
        result[disk] = { read_bytes: 0, write_bytes: 0 }
      end
    end
    result
  end

  def calculate_speeds(prev, curr)
    speeds = {}
    curr.each do |disk, data|
      if prev[disk]
        read_mb = (data[:read_bytes] - prev[disk][:read_bytes]) / (1024.0 * 1024.0)
        speeds[disk] = read_mb / @interval
      end
    end
    speeds
  end

  def collect
    puts "\e[36m💾 Disk Monitor (History) (Ruby)\e[0m"
    puts "Интервал: #{@interval} сек, Длительность: #{@duration} сек"
    puts

    total_samples = @duration / @interval
    puts "📊 Мониторинг дисков..."

    prev = get_disk_io
    sleep @interval

    total_samples.times do |i|
      curr = get_disk_io
      speeds = calculate_speeds(prev, curr)

      entry = {
        timestamp: Time.now.iso8601,
        speeds: speeds
      }
      @history << entry

      # Прогресс
      percent = (i + 1).to_f / total_samples * 100
      bar_len = 30
      filled = (bar_len * (i + 1) / total_samples).to_i
      bar = '█' * filled + '░' * (bar_len - filled)
      STDERR.print "\r[#{bar}] #{'%.1f' % percent}% (#{i+1}/#{total_samples})"

      prev = curr
      sleep @interval
    end

    STDERR.puts
    save_history
    print_stats
  end

  def save_history
    data = {
      interval: @interval,
      duration: @duration,
      start_time: @history.first&.[](:timestamp),
      end_time: @history.last&.[](:timestamp),
      samples: @history
    }
    File.write(@history_file, JSON.pretty_generate(data))
    puts "\e[32m💾 История сохранена: #{@history_file}\e[0m"
  end

  def print_stats
    return if @history.empty?

    stats = Hash.new { |h, k| h[k] = [] }
    @history.each do |entry|
      entry[:speeds].each do |disk, speed|
        stats[disk] << speed
      end
    end

    puts "\n\e[36m📈 Статистика за период:\e[0m"
    puts "┌" + "─" * 80 + "┐"
    puts "│ #{'Диск'.ljust(10)} #{'Среднее (МБ/с)'.ljust(15)} #{'Макс (МБ/с)'.ljust(12)} #{'Мин (МБ/с)'.ljust(12)} #{'IOPS'.ljust(15)} │"
    puts "├" + "─" * 80 + "┤"

    stats.each do |disk, speeds|
      if speeds.any?
        avg = speeds.sum / speeds.size
        max = speeds.max
        min = speeds.min
        avg_iops = avg
        puts "│ #{disk.ljust(10)} #{'%12.1f' % avg} #{'%12.1f' % max} #{'%12.1f' % min} #{'%12.0f' % avg_iops}  │"
      end
    end

    puts "└" + "─" * 80 + "┘"
  end

  def load_history
    unless File.exist?(@history_file)
      puts "\e[31m❌ Файл #{@history_file} не найден.\e[0m"
      return
    end

    data = JSON.parse(File.read(@history_file), symbolize_names: true)
    puts "\e[36m📂 Загружена история из #{@history_file}\e[0m"
    puts "  Интервал: #{data[:interval]} сек"
    puts "  Начало: #{data[:start_time]}"
    puts "  Конец: #{data[:end_time]}"
    puts "  Сэмплов: #{data[:samples]&.size || 0}"
  end
end

def main
  options = { interval: 2, duration: 60, history_file: 'disk_history.json', view: false }
  OptionParser.new do |opts|
    opts.banner = "Usage: ruby disk_monitor.rb [options]"
    opts.on("--interval N", Integer, "Интервал сбора") { |v| options[:interval] = v }
    opts.on("--duration N", Integer, "Длительность сбора") { |v| options[:duration] = v }
    opts.on("--file FILE", "Файл истории") { |v| options[:history_file] = v }
    opts.on("--view", "Просмотреть историю") { options[:view] = true }
  end.parse!

  monitor = DiskMonitor.new(**options)

  if options[:view]
    monitor.load_history
  else
    monitor.collect
  end
end

main if __FILE__ == $0
